### PR TITLE
Support JSON as source for parametrized tests arguments

### PR DESCRIPTION
**Pull request by filiphr**
This is an initial implementation fro #101. It currently only supports [Jackson 2](https://github.com/FasterXML/jackson-databind) for Json Parsing. However, it should be fairly simple to add support for different parsers.

I know the policy about no runtime dependencies. Therefore, I tried using [Optional Dependencies](https://blog.gradle.org/optional-dependencies) from Gradle and its [feature variants](https://docs.gradle.org/6.0.1/userguide/feature_variants.html). If this is something interesting we can look into doing this with different source sets in order for complete isolation (currently the Checkstyle import-control is handling this).

There is still documentation missing. However, if you like the approach I will work on the documentation.

Proposed commit message:

```
Support for JSON as source for parametrized tests

Closes: #101
PR: #492
```

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [x] There is documentation (Javadoc and site documentation; added or updated)
* [x] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [x] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [x] Only one sentence per line (especially in `.adoc` files)
* [x] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [x] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [x] The `package-info.java` contains information about the new extension

Code
* [x] Code adheres to code style, naming conventions etc.
* [x] Successful tests cover all changes
* [x] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [x] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [x] A prepared commit message exists
* [x] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

**Comment by Bukama** [comment]
I would like to pass this to the other maintainers cause of the dependency thing and changes in the build script

File: src/main/java/org/junitpioneer/jupiter/json/JsonSource.java
Comment (Michael1993) [inline src/main/java/org/junitpioneer/jupiter/json/JsonSource.java:48]: This is confusing. JUnit 5 has `@CsvSource` which can take CSV as a `String` and `CsvFileSource` which takes CSV file names. I think following that, this should be `JsonFileSource`.

Comment (filiphr) [reply src/main/java/org/junitpioneer/jupiter/json/JsonSource.java:48]: You are right about this. It would make more sense this to be named `@JsonFileSource`. I didn't really think about the naming and took what was there in the issue

**Review by Michael1993 — CHANGES_REQUESTED** [review]
Pioneer actually has a `module-info.java` and a modular build, so you need to update that if you want the build to succeed. You can find it at `src/main/java/module/module-info.java`.

> (package com.fasterxml.jackson.databind is declared in module com.fasterxml.jackson.databind, but module org.junitpioneer does not read it)

**Comment by nipafx** [comment]
From just looking at the tests, this seems to be exactly what we had in mind. Good job! I think it makes sense to pursue this further. 👍🏾

File: src/main/java/org/junitpioneer/jupiter/json/JsonSource.java
Comment (filiphr) [reply src/main/java/org/junitpioneer/jupiter/json/JsonSource.java]: I renamed this into `@JsonFileSource`.

**Comment by filiphr** [comment]
>Pioneer actually has a `module-info.java` and a modular build, so you need to update that if you want the build to succeed. You can find it at `src/main/java/module/module-info.java`.

Thanks @Michael1993, I've updated the file. The modular build is now working correctly.

I've taken this a bit further, added some documentation, and did some improvements;

* Support for using `@Param` for single method arguments, by extracting only one specific element from it
* Support for deconstructing into complex types

**Comment by nipafx** [comment]
Thank you for this PR, @filiphr. You've seen my streams, you know how much time management challenges me. I will take a closer look at this, but it will take some more time before I do.

Also, we need to resolve #502 before we can merge/move this.

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:11]:
```suggestion
Depending on the test method parameters, the extraction of the values might differ.
```

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:15]:
```suggestion
If the method has a single argument, the json object argument will be converted to that type.
```

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:55]: I appreciate the attention to detail, but I don't think we need to flesh out the test. 😉

File: src/main/java/org/junitpioneer/jupiter/json/Param.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/json/Param.java:28]: This name should probably be a bit longer - see [related discussion](https://github.com/junit-pioneer/junit-pioneer/issues/415#issuecomment-861770281) on Cartesian tests.

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:83]:
```suggestion
If the method has multiple arguments, each json object argument will be deconstructed to each of the method arguments.
```

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:104]: As above.

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:149]: As above.

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:156]: Can you add a section on thread safety like the other extensions have? Will probably look [like this](https://junit-pioneer.org/docs/range-sources/#thread-safety).

File: src/main/java/org/junitpioneer/jupiter/json/JacksonJsonConverter.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/json/JacksonJsonConverter.java:23]: Can you explain why we need this class and `JacksonNode`?

Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/json/JacksonJsonConverter.java:23]: I also don't know why we need `JsonConverter` and `JsonConverterProvider`. 🙈 Halp?

File: src/main/java/org/junitpioneer/jupiter/json/JsonFileArgumentsProvider.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/json/JsonFileArgumentsProvider.java:38]: Since we're rebuilding Cartesian product tests, I'm not sure whether it makes sense to implement `CartesianAnnotationConsumer<JsonFileSource>`. @Michael1993, what do you think?

File: src/test/java/org/junitpioneer/jupiter/json/JsonFileSourceArgumentsProviderTests.java
Comment (nipafx) [inline src/test/java/org/junitpioneer/jupiter/json/JsonFileSourceArgumentsProviderTests.java:67]: Good job with these tests! Since they already check all the correct values, don't you think we can remove the assertions in the actual tests below? Otherwise, all changes need to be made in two places.

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:113]:
```suggestion
== Extracting nested array
```

Shouldn't this be h2 - it applies to single-argument methods as well, right?

File: docs/json.adoc
Comment (nipafx) [inline docs/json.adoc:4]: I propose to slightly change the doc's structure. Add another sentence or two here what to look out for (single vs multiple arguments; nested arrays) and then add `== Method Arguments` and make the next two headers on single and multiple argument methods h3.

**Review by nipafx — CHANGES_REQUESTED** [review]
This looks really good, I think we can merge it "soon" (modulo my very long latency). A few comments below.

Also, if I'm not missing something, this "only" addresses part of #101. The `ArgumentConverter` and a source that parses inline JSON (as opposed to from a file) are also mentioned there. If we close #101, we need to open new issues for those. (If we still want to have them.)

File: src/main/java/org/junitpioneer/jupiter/json/JacksonJsonConverter.java
Comment (filiphr) [reply src/main/java/org/junitpioneer/jupiter/json/JacksonJsonConverter.java:23]: The idea of the `JsonConverter` and `Node` is to provide an extra layer of abstraction over the Json Parsers. This would allow for using different parser implementations, such as [GSON](https://github.com/google/gson), [Johnzon](https://johnzon.apache.org/), [Moshi](https://github.com/square/moshi), without any changes in the main logic. 

This was inspired by the [`NodeFactory`](https://github.com/lukas-krecan/JsonUnit/blob/master/json-unit-core/src/main/java/net/javacrumbs/jsonunit/core/internal/NodeFactory.java) from the JsonUnit project.

See also `JsonConverterProvider` in this PR.

I will add some Javadoc to `JsonConverter` and `Node` to make it clearer.

File: src/main/java/org/junitpioneer/jupiter/json/JsonFileArgumentsProvider.java
Comment (filiphr) [reply src/main/java/org/junitpioneer/jupiter/json/JsonFileArgumentsProvider.java:38]: I've added this because I saw it in the range functionality. I can remove it if it is not needed.

File: src/test/java/org/junitpioneer/jupiter/json/JsonFileSourceArgumentsProviderTests.java
Comment (filiphr) [reply src/test/java/org/junitpioneer/jupiter/json/JsonFileSourceArgumentsProviderTests.java:67]: We can remove them yes, but then if you want to debug you'll need to write something to test things out.

If new things are added I expect new json files to be used.

File: docs/json.adoc
Comment (filiphr) [reply docs/json.adoc:4]: Will doo that

File: docs/json.adoc
Comment (filiphr) [reply docs/json.adoc:156]: Makes sense. I'll do that

File: src/main/java/org/junitpioneer/jupiter/json/Param.java
Comment (filiphr) [reply src/main/java/org/junitpioneer/jupiter/json/Param.java:28]: I will change it to `Parameter`. If you have better idea let me know

File: docs/json.adoc
Comment (filiphr) [reply docs/json.adoc:55]: 😄  I'll remove them from the docs

**Comment by filiphr** [comment]
>This looks really good, I think we can merge it "soon" (module my very long latency).

Thanks for the review. I've applied the feedback and have answered some of your comments. I am not in a rush (as you can see I have a long latency as well). 

>Also, if I'm not missing something, this "only" addresses part of #101. The ArgumentConverter and a source that parses inline JSON (as opposed to from a file) are also mentioned there. If we close #101, we need to open new issues for those. (If we still want to have them.)

Indeed this solves only a part of #101, the one where data is read from a classpath resource or a file. Reading from an inline JSON would be quite trivial as well. I'll try to add it as part of this PR as well.

I am not sure about the `ArgumentConverter` though, what would you expect from it? Currently we extract the properties in the types that they are in the JSON (string, number, boolean, etc.)

**Comment by filiphr** [comment]
By the way, I forgot to mention, I watched the stream afterwards to see your thought process when reviewing this 😄 . Couldn't be there when you were doing it live

File: src/main/java/org/junitpioneer/jupiter/json/JsonFileArgumentsProvider.java
Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/json/JsonFileArgumentsProvider.java:37]: Ad hoc decision: For new extensions that can interact with Cartesian tests, let's only implement the new approach. Can you change this? :)

File: src/test/java/org/junitpioneer/jupiter/json/JsonFileSourceArgumentsProviderTests.java
Comment (nipafx) [reply src/test/java/org/junitpioneer/jupiter/json/JsonFileSourceArgumentsProviderTests.java]: Got it. It's unfortunate, but makes sense.

File: src/main/java/org/junitpioneer/jupiter/json/JacksonJsonConverter.java
Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/json/JacksonJsonConverter.java:24]: Got it!

**Review by nipafx — CHANGES_REQUESTED** [review]
Looking good, but we found some things still left to do:

* whatever we decide [on the issue](https://github.com/junit-pioneer/junit-pioneer/issues/101#issuecomment-968281501)
* support new Cartesian tests instead of the old approach
* I still need to wrap my head around #502 (this PR currently uses [Gradle's feature variants](https://docs.gradle.org/current/userguide/feature_variants.html)).

File: src/main/java/org/junitpioneer/jupiter/json/JsonConverterProvider.java
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/json/JsonConverterProvider.java:35]: This error case comes up when users haven't configured a (supported) JSON parse, so we need to make sure, it's visible. Both Maven and Gradle will cut the error message very short, so it's unclear what the problem is.

Some options:

* throw `NoJsonParserConfiguredException` (extends `PreconditionViolationException`)
* print to `System.err`
* log a report entry with JUnit Jupiter

We should also mention which parsers we support.

Comment (filiphr) [reply src/main/java/org/junitpioneer/jupiter/json/JsonConverterProvider.java:35]: Which Maven did you use? Using 3.8.4 I get

>[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.031 s <<< FAILURE! - in org.junitpioneer.demo.JsonSourceTests
[ERROR] org.junitpioneer.demo.JsonSourceTests.jsonTest(String, int)  Time elapsed: 0.001 s  <<< ERROR!
org.junit.platform.commons.PreconditionViolationException: There is no available Json parsing library

Nothing is cut short. I will add more information about which json parsers are currently supported. 

For gradle when I do `./gradlew test` I only get that test failed. In the test report you can see the full error with its stacktrace.

When I add

```
tasks.named<Test>("test") {
	useJUnitPlatform()

	testLogging {
		setExceptionFormat("full")
	}
}
```

then I get

>JsonSourceTests > jsonTest(String, int) > org.junitpioneer.demo.JsonSourceTests.initializationError FAILED
    org.junit.platform.commons.PreconditionViolationException: There is no available Json parsing library
        at app//org.junitpioneer.jupiter.json.JsonConverterProvider.getJsonConverter(JsonConverterProvider.java:35)
        .... There is actually more stacktrace


Regarding the options you mentioned:

* What is the benefit of throwing `NoJsonParserConfiguredException`?
* I am not a fan of printing to `System.err`. So will leave it to you to add it in case you really really want it
* I can log a report entry with JUnit Jupiter, need to see how to do that first.

Comment (nipafx) [reply src/main/java/org/junitpioneer/jupiter/json/JsonConverterProvider.java:35]: I don't recall the details any more (stupid me, I should've added what I saw as text or screenshot), so I'll rely on your analysis. Regarding the options:

* The exception name showed up prominently - if that's `NoJsonParserConfiguredException`, it might not even be necessary to read the text.
* Me neither, let's not do it then.
* you know what, ignore that. I think it makes sense to limit report entries to (mostly) successful operations, not for situations that abort the test run.

File: src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTests.java
Comment (nipafx) [inline src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTests.java:341]: Why did you rename this? It looks like it's indeed related to parameterized tests.

File: docs/json-argument-source.adoc
Comment (nipafx) [inline docs/json-argument-source.adoc:1]: I renamed the file (and thus the URL) and changed the title to limit it to JSON argument sources - who knows what other fancy things we can do with JSON in the future that would otherwise name-clash. 😉

File: src/test/java/org/junitpioneer/jupiter/json/JsonClasspathSourceArgumentsProviderTests.java
Comment (nipafx) [inline src/test/java/org/junitpioneer/jupiter/json/JsonClasspathSourceArgumentsProviderTests.java:37]: No tests for files? Too brittle?

**Review by nipafx — CHANGES_REQUESTED** [review]
Looks really good. The documentation doesn't reflect the three annotations yet (I only updated the intro up until the bullet point list), but otherwise this looks good to merge. 👍🏾

**Comment by nipafx** [comment]
I just noticed that the module declaration (in `src/modules`) still needs to be updated.

File: src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTests.java
Comment (filiphr) [reply src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTests.java:341]: IntelliJ hickup during some other renaming. I'll revert it

**Comment by filiphr** [comment]
I closed this by accident. I've added some replies to your comments. And I fixed some of the problems that you found. Let me know what you think.

I've also added myself to the contributors list to this list, I hope I didn't mess this up.

**Comment by nipafx** [comment]
> What is missing in the documentation?

It still uses the old-style `@JsonFileSource(resources = "/jedis.json")` syntax. While udpating that, you could also create the snippets as a proper demo in `src/demo/java`, which gets compiled, so we see when things are outdated. See `report-entries.adoc` for how to reference them (note the attribute `:xp-demo-dir:` at the top).

> I've also added myself to the contributors list to this list, I hope I didn't mess this up.

Nope, looks good. And you definitely belong there for this 12-month-juggernaut. :D

> What needs to be updated there?

Nothing, looks good. 😬

File: src/test/java/org/junitpioneer/jupiter/json/JsonClasspathSourceArgumentsProviderTests.java
Comment (nipafx) [reply src/test/java/org/junitpioneer/jupiter/json/JsonClasspathSourceArgumentsProviderTests.java:37]: This should work, right?

```java
@ParameterizedTest
@JsonFileSource("build/resources/test/org/junitpioneer/jupiter/json/jedis.json")
void deconstructCustomerMultipleLinesComplexType(@Property("name") String name, @Property("height") int height) {
	assertThat(Collections.singleton(tuple(name, height)))
		.containsAnyOf(tuple("Luke", 172), tuple("Yoda", 66));
}
```

**Comment by filiphr** [comment]
I've updated the documentation. It was a good thing to mention the use of proper compilable demo, because even the test annotation had a typo. I think everything is covered now, if there is something more missing, please do let me know

**Comment by nipafx** [comment]
Commit message:

```
Support JSON sources for parametrized tests 

This extension needs a JSON parser. In order not do pull in a
specific run-time dependency, it makes use of Gradle's feature
variants as described in CONTRIBUTING.md (including Checkstyle
rules). For now, only Jackson is supported as JSON parser. Other
parsers can be added in future versions if users need them.

Closes: #101
PR: #492
```

**Review by nipafx — APPROVED** [review]

**Comment by nipafx** [comment]
Great job, @filiphr, and big thanks for your patience!

**Comment by filiphr** [comment]
Thanks for merging this @nipafx. It was fun working on this, it doesn't matter how long it took :)

**Pull request merged**
